### PR TITLE
config: respect proxy_* settings in agent

### LINKED ISSUE #245: APM Trace agent doesn't use http proxy configuration for reporting
I'm attempting to use APM in an environment that restricts outbound http unless through an http proxy, and from what I can tell this agent is not supporting the http proxy config (or I'm missing the correct place to put it).

I do have at present (`proxy_host`, `proxy_port`, `proxy_user`, `proxy_password`) set inside `/etc/dd-agent/datadog.conf` and have restarted the datadog-agent.

I see the following in my logs:

```
tail -f /var/log/datadog/trace-agent.log
2017-03-16 14:21:50 ERROR (endpoint.go:119) - error when requesting to endpoint https://trace.agent.datadoghq.com/api/v0.1/collector: Post https://trace.agent.datadoghq.com/api/v0.1/collector?api_key=[redacted]: dial tcp 23.23.189.219:443: i/o timeout
2017-03-16 14:21:52 INFO (receiver.go:329) - receiver handled 51 spans, dropped 0 ; handled 51 traces, dropped 0
2017-03-16 14:22:06 ERROR (endpoint.go:119) - error when requesting to endpoint https://trace.agent.datadoghq.com/api/v0.1/collector: Post https://trace.agent.datadoghq.com/api/v0.1/collector?api_key=[redacted]: dial tcp 50.16.188.192:443: i/o timeout
```

### PR DESCRIPTION
**Pull request by talwai**
when [proxy
settings](https://github.com/DataDog/dd-agent/blob/master/datadog.conf.example#L7-L10) are defined in datadog.conf
update the default client in the writer to submit to a proxy.

fixes #245

### CODE REVIEW

File: config/proxy.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	return &p
+}
+
+// URL turns ProxySettings into an idiomatic URL struct
+func (p *ProxySettings) URL() (*url.URL, error) {
+	// construct scheme://user:pass@host:port
+	var userpass string
+	if p.User != "" {
+		if p.Password != "" {
+			userpass = fmt.Sprintf("%s:%s@", p.User, p.Password)
```
Comment (ufoot) [inline config/proxy.go]: I think we could URL encode here (cf https://www.cyberciti.biz/faq/unix-linux-export-variable-http_proxy-with-special-characters/ or http://unix.stackexchange.com/questions/42273/export-http-proxy-and-special-characters-in-passwd ). Probably using https://golang.org/pkg/net/url/#UserPassword would do the job.

File: config/proxy.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package config
```
Comment (ufoot) [inline config/proxy.go]: I like the idea of having a separate chunk of code / struct to handle this. Indeed, it is not really agent specific.

File: config/proxy_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	s, err = getURL(f)
+	assert.Nil(err)
+	assert.Equal("http://aaditya:password_12@myhost:3129", s.String())
+
+	f, _ = ini.Load([]byte(strings.Join([]string{
+		"[Main]",
+		"proxy_host = https://myhost",
+		"proxy_port = 3129",
+		"proxy_user = aaditya",
+		"proxy_password = password_12",
```
Comment (ufoot) [inline config/proxy_test.go]: You should probably add a test with some special chars such as `/:\!?&;=@éÔγλῶσσα`, and test that `ProxySettings.URL()` returns the expected value.

**Comment by talwai** [comment]
Yes good catch on the special chars handling. I've addressed this in https://github.com/DataDog/datadog-trace-agent/commit/9486db3

**Pull request merged**